lib/vrl/stdlib/src/for_keys.rs

```diff
@@ -0,0 +1,195 @@
+        }
+    }
+
+    fn new_from_parent(parent: Vec<String>, values: btree_map::Iter<'a, String, Value>) -> Self {
+        Self {
+            values,
+            inner: None,
+            parent: Some(parent),
+        }
+    }
+
+    fn new_key(&self, key: &str) -> Vec<String> {
+        match &self.parent {
+            None => vec![key.to_string()],
+            Some(parent) => {
+                let mut copy = parent.clone();
+                copy.push(key.to_string());
+                copy
+            }
+        }
+    }
+}
+
+impl<'a> std::iter::Iterator for MapParent<'a> {
+    type Item = (Vec<String>, &'a Value);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(ref mut inner) = self.inner {
+            let next = inner.next();
+            match next {
+                Some(_) => return next,
+                None => self.inner = None,
+            }
+        }
+
+        let next = self.values.next();
+        match next {
+            Some((key, Value::Object(value))) => {
+                self.inner = Some(Box::new(MapParent::new_from_parent(
+                    self.new_key(key),
+                    value.iter(),
+                )));
+                self.next()
+            }
+            Some((key, value)) => Some((self.new_key(key), value)),
+            None => None,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct ForKeys;
+
+impl Function for ForKeys {
+    fn identifier(&self) -> &'static str {
+        "for_keys"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::OBJECT,
+            required: true,
+        }]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "iterate object key values",
+            source: r#".example = {}; .example.child = "123"; for_keys(.) -> |keys, value| { if join!(keys, ".") == "example.child" { . = remove!(., keys) }}; ."#,
+            result: Ok(r#"{"example":{}}"#),
+        }]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let closure = arguments.required_closure()?;
+
+        Ok(ForKeysFn { value, closure }.as_expr())
+    }
+
+    fn closure(&self) -> Option<closure::Definition> {
+        use closure::{Definition, Input, Output, Variable, VariableKind};
+
+        Some(Definition {
+            inputs: vec![Input {
+                parameter_keyword: "value",
+                kind: Kind::object(Collection::any()).or_array(Collection::any()),
+                variables: vec![
+                    Variable {
+                        kind: VariableKind::Exact(Kind::array(Collection::any())),
+                    },
+                    Variable {
+                        kind: VariableKind::TargetInnerValue,
```

Comment:
Regarding adding support for arrays - how'd you want this function to handle them?

My thoughts would either be:
- [Option 1] Calling `for_each_recursive` with an _ARRAY_ it will recursively iterate through any arrays in the array. Providing an array of indices (eg. [1, 4, 2]) and the value. Calling it with an _OBJECT_, it will only recursively iterate through the object keys, arrays will not be iterated.
- [Option 2] Alternatively calling `for_each_recursive` with an _ARRAY_ or _OBJECT_ the function will iterate through the object/arrays, providing a "path" to the value with a mix of types (eg. [2, "example", 3, "stringkey"]) and the value. _Everything_ in whatever type is passed to the function will be iterated through.

Option 1 suites our use case for this function, while option 2 would be less ideal for us if using it for schema/field validation (but could be worked around). 

Let me know your opinion/thoughts 😃

Reply:
Option 1 is fine. We should use the same behavior as the existing `for_each` function, and that's what it does currently.